src/main/java/uk/gov/justice/hmpps/prison/service/AgencyServiceImpl.java

```diff
@@ -255,23 +257,32 @@ private boolean isBlankAddress(final PrisonContactDetail pcd) {
     public List<OffenderCell> getCellsWithCapacityInAgency(@NotNull final String agencyId, String attribute) {
         final var livingUnits = livingUnitRepository.findAllByAgencyLocationId(agencyId);
         return livingUnits.stream()
-                .filter(LivingUnit::isActiveCellWithSpace)
-                .map(livingUnit -> OffenderCell.builder()
-                    .capacity(livingUnit.getOperationalCapacity())
-                    .noOfOccupants(livingUnit.getNoOfOccupants())
-                    .id(livingUnit.getLivingUnitId())
-                    .description(livingUnit.getUserDescription() != null ?  livingUnit.getUserDescription() : livingUnit.getDescription())
-                    .attributes(livingUnitProfileRepository
-                                .findAllByLivingUnitIdAndAgencyLocationIdAndDescription(livingUnit.getLivingUnitId(), livingUnit.getAgencyLocationId(), livingUnit.getDescription())
-                                .stream()
-                                .filter(LivingUnitProfile::isAttribute)
-                                .map(profile -> OffenderCellAttribute.builder()
-                                     .code(profile.getHousingAttributeReferenceCode().getCode())
-                                     .description(profile.getHousingAttributeReferenceCode().getDescription())
-                                     .build())
-                                .collect(Collectors.toList()))
-                    .build())
-                .filter(cell -> attribute == null || cell.getAttributes().stream().map(OffenderCellAttribute::getCode).collect(Collectors.toList()).contains(attribute))
-                .collect(Collectors.toList());
+                .filter(livingUnit -> {
+                    final var agencyLocation = agencyInternalLocationRepository.findOneByDescription(livingUnit.getDescription());
```

Comment:
Rather than calling `agencyInternalLocationRepository.findOneByDescription(livingUnit.getDescription())` twice (once to filter, then again in the map), you could do something like:
```
return livingUnits.stream()
            .map(this::transform)
            .flatMap(Optionals::stream)
            .filter(cell -> attribute == null || cell.getAttributes().stream().map(OffenderCellAttribute::getCode).collect(toList()).contains(attribute))
            .collect(toList());
```
where `transform` is a method that returns an optional like:

```
private Optional<OffenderCell> transform(final LivingUnit livingUnit) {
    return agencyInternalLocationRepository.findOneByDescription(livingUnit.getDescription())
        .filter(AgencyInternalLocation::isActiveCellWithSpace)
        .map(agencyLocation -> 
              OffenderCell.builder()
                      ...
              .build()
        );
}
```

Reply:
Cheers! I am very sure I will back looking at this endpoint soon so will do this then :)